homeassistant/components/tasmota/__init__.py

```diff
@@ -72,33 +73,48 @@ async def _unsubscribe_topics(sub_state: dict | None) -> dict:
 
     tasmota_mqtt = TasmotaMQTTClient(_publish, _subscribe_topics, _unsubscribe_topics)
 
-    device_registry = await hass.helpers.device_registry.async_get_registry()
+    device_registry = dr.async_get(hass)
 
     async def async_discover_device(config: TasmotaDeviceConfig, mac: str) -> None:
         """Discover and add a Tasmota device."""
         await async_setup_device(
             hass, mac, config, entry, tasmota_mqtt, device_registry
         )
 
-    async def async_device_removed(event: Event) -> None:
+    async def async_device_updated(event: Event) -> None:
         """Handle the removal of a device."""
-        device_registry = await hass.helpers.device_registry.async_get_registry()
-        if event.data["action"] != "remove":
+        device_registry = dr.async_get(hass)
+        device_id = event.data["device_id"]
+        if event.data["action"] not in ("remove", "update"):
             return
 
-        device = device_registry.deleted_devices[event.data["device_id"]]
-
-        if entry.entry_id not in device.config_entries:
-            return
-
-        macs = [c[1] for c in device.connections if c[0] == CONNECTION_NETWORK_MAC]
+        connections: set[tuple[str, str]]
+        if event.data["action"] == "update":
+            if "config_entries" not in event.data["changes"]:
+                return
+
+            device = device_registry.async_get(device_id)
+            if not device:
+                # The device is already removed, do cleanup when we get "remove" event
+                return
+            if entry.entry_id in device.config_entries:
+                # Not removed from device
+                return
+            connections = device.connections
+        else:
+            deleted_device = device_registry.deleted_devices[event.data["device_id"]]
+            connections = deleted_device.connections
+            if entry.entry_id not in deleted_device.config_entries:
+                return
+
+        macs = [c[1] for c in connections if c[0] == CONNECTION_NETWORK_MAC]
```

Comment:
Why not wait until we have `async_remove_config_entry_device` active. Then we can access connections on the device entry passed to `async_remove_config_entry_device`. We won't need the code above, right?